plugins/usbdmx/JaRuleWidget.cpp

```diff
@@ -18,50 +18,253 @@
+// ----------------------------------------------------------------------------
+
+bool JaRuleWidget::InternalInit() {
+  struct libusb_config_descriptor *config;
+  int error = m_adaptor->GetActiveConfigDescriptor(m_device, &config);
+  if (error) {
+    OLA_WARN << "Failed to get active config descriptor for Ja Rule device: "
+             << m_adaptor->ErrorCodeToString(error);
+    return false;
+  }
+
+  // Each endpoint address is 8 bits. Bit 7 is the endpoint direction (in/out).
+  // The lower 4 bits are the endpoint number. We try to find bulk endpoints
+  // with matching numbers.
+  typedef std::map<uint8_t, EndpointCapabilties> EndpointMap;
+  EndpointMap endpoint_map;
+
+  for (uint8_t iface_index = 0; iface_index < config->bNumInterfaces;
+       iface_index++) {
+    const struct libusb_interface &iface = config->interface[iface_index];
+    // We don't support alt settings.
+    if (iface.num_altsetting != 1) {
+      continue;
+    }
+    const struct libusb_interface_descriptor &iface_descriptor =
+        iface.altsetting[0];
+    if (iface_descriptor.bInterfaceClass == LIBUSB_CLASS_VENDOR_SPEC &&
+        iface_descriptor.bInterfaceSubClass == 0xff &&
+        iface_descriptor.bInterfaceProtocol == 0xff) {
+      // Vendor class, subclass & protocol
+      for (uint8_t endpoint_index = 0; endpoint_index <
+           iface_descriptor.bNumEndpoints; endpoint_index++) {
+        const struct libusb_endpoint_descriptor &endpoint =
+            iface_descriptor.endpoint[endpoint_index];
+        if ((endpoint.bmAttributes & LIBUSB_TRANSFER_TYPE_MASK) !=
+             LIBUSB_TRANSFER_TYPE_BULK) {
+          continue;
+        }
+
+        uint8_t endpoint_address = endpoint.bEndpointAddress;
+        uint8_t endpoint_number = (
+            endpoint_address & LIBUSB_ENDPOINT_ADDRESS_MASK);
+        uint8_t endpoint_direction = (
+            endpoint_address & LIBUSB_ENDPOINT_DIR_MASK);
+
+        if (endpoint_direction == LIBUSB_ENDPOINT_IN) {
+          endpoint_map[endpoint_number].in_supported = true;
+          endpoint_map[endpoint_number].in_interface = iface_index;
+        }
+        if (endpoint_direction == LIBUSB_ENDPOINT_OUT) {
+          endpoint_map[endpoint_number].out_supported = true;
+          endpoint_map[endpoint_number].out_interface = iface_index;
+        }
+      }
+    }
+  }
+
+  m_adaptor->FreeConfigDescriptor(config);
+
+  if (!m_adaptor->OpenDevice(m_device, &m_usb_handle)) {
+    return false;
+  }
+
+  std::set<int> interfaces_to_claim;
+
+  EndpointMap::const_iterator endpoint_iter = endpoint_map.begin();
+  uint8_t port_index = 0;
+  for (; endpoint_iter != endpoint_map.end(); ++endpoint_iter) {
+    const EndpointCapabilties &capabilites = endpoint_iter->second;
+    if (capabilites.in_supported && capabilites.out_supported) {
+      interfaces_to_claim.insert(capabilites.in_interface);
+      interfaces_to_claim.insert(capabilites.out_interface);
+      OLA_INFO << "Found Ja Rule port at "
```

Comment:
Perhaps it should be Ja Rule when not in code; my brother got the reference by the way.